core/server/common/src/main/java/alluxio/master/transport/Listeners.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.transport;
+
+import com.google.common.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.function.Consumer;
+
+/**
+ * A class to hold all listeners belong to given type T.
+ *
+ * @param <T> the listener type
+ */
+public class Listeners<T> implements Iterable<Listener<T>> {
```

Comment:
Is there any Java built-in or guava alternative of this listener we can use?

Reply:
The current implementation bounds the listeners with a shared thread context. So that all the listeners accept method will be executed by the same executor.